cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Apereo under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work
+ * for additional information regarding copyright ownership.
+ * Apereo licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License.  You may obtain a
+ * copy of the License at the following location:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.jasig.cas.support.pac4j.authentication.handler.support;
+
+import io.cos.cas.adaptors.postgres.handlers.OpenScienceFrameworkInstitutionHandler;
+import org.apache.commons.lang3.StringUtils;
+import org.jasig.cas.authentication.BasicCredentialMetaData;
+import org.jasig.cas.authentication.DefaultHandlerResult;
+import org.jasig.cas.authentication.HandlerResult;
+import org.jasig.cas.authentication.PreventedException;
+import org.jasig.cas.support.pac4j.authentication.principal.ClientCredential;
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.profile.UserProfile;
+
+import javax.security.auth.login.FailedLoginException;
+import javax.validation.constraints.NotNull;
+import java.security.GeneralSecurityException;
+
+/**
+ * Specialized handler which builds the authenticated user directly from the retrieved user profile.
+ *
+ * @author Jerome Leleu
+ * @author Longze Chen
+ * @since 4.1.0
+ */
+public class ClientAuthenticationHandler extends AbstractClientAuthenticationHandler {
+
+    /** Whether to use the typed identifier (by default) or just the identifier. */
+    private boolean typedIdUsed = true;
+
+    /** The handler for osf institution. */
+    @NotNull
+    private final OpenScienceFrameworkInstitutionHandler institutionHandler;
```

Comment:
Need an instance of `OpenScienceFrameworkInstitutionHandler` to access ` OpenScienceFrameworkInstitution` model and dao.